xrdsst/controllers/base.py

```diff
@@ -59,16 +59,19 @@ def _pre_argument_parsing(self):
                                 metavar='file',
                                 default=os.path.join(ROOT_DIR, BaseController._DEFAULT_CONFIG_FILE))
 
-    def create_api_key(self, roles_list, config, security_server):
+    def create_api_key(self, roles_list, security_server):
```

Comment:
Fixed general exception in create_api_key method to have more information related to exit_code and output data.

Reply:
This now indeed passes the Exception message forward, but with the previously cited configuration, where just ``admin_credentials`` is moved to root leve, it now fails as follows:

```sh
$ xrdsst -c config/newss5.yaml token list
Exception calling BaseController->get_api_key:: 'NoneType' object is not iterable
Exception when calling TokensApi->get_tokens: (401)
Reason: 
HTTP response headers: HTTPHeaderDict({'x-road-ui-correlation-id': 'c89999342193ee17', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': '0', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'none'", 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Date': 'Wed, 31 Mar 2021 08:10:02 GMT'})
HTTP response body: {"status":401}

```

Reply:
The current implementation required the admin_credentials, ssh_private_key and ssh_user paramters to still be present in the configuration, but could be left empty, the exception resulted when these were totally missing, now amended the source code to check whether these are totally missing

Reply:
Yup, previously presented configuration snippet now works at console as expected, to have this staked down, add good testcase that tests configuration propagation from configuration root to security servers and their priority, asserting that values that end up being used for ``ssh`` are the same as expected/documented desired.

Reply:
Added 6 unit tests for testing retrieval of admin_credentials and SSH parameters from different configuration file sections

Reply:
Tests seem okay at first glance.

Further manual testing that the configured key is not the one solely used, e.g. creating brand new keypair ``justnewkey`` and robustly substituting it into configuration file:

```diff
-private_key: /home/user/.ssh/real
+private_key: /home/user/.ssh/justnewkey
```

still allows API key creation, if only few keys are present under ``~/.ssh``, since SSH makes attempts to use other keys under  there, which can either succeed when the amount of keys are small or fails with ``Too many authentication failures`` when there are more keys and the one that allows access to server is not reached 'fast enough'.

It seems it is better to fail at once there, instead of attempting fallbacks, so that misconfigurations are revealed earlier.